security/advancedtls/provider.go

```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/grpclog"
+)
+
+const (
+	defaultInterval = 1 * time.Hour
+)
+
+var logger = grpclog.Component("channelz")
+
+// IdentityPemFileProviderOptions contains fields to be filled out by users
+// for obtaining identity private key and certificates from PEM files.
+type IdentityPemFileProviderOptions struct {
+	CertFile string
+	KeyFile  string
+	Interval time.Duration
+}
+
+// IdentityPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date identity private key
+// and certificates based on the input PEM files.
+type IdentityPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// RootPemFileProviderOptions contains fields to be filled out by users
+// for obtaining root certificates from PEM files.
+type RootPemFileProviderOptions struct {
+	TrustFile string
+	Interval  time.Duration
+}
+
+// RootPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date root certificates based on the input PEM files.
+type RootPemFileProvider struct {
```

Comment:
My thought is that it allows users to have the flexibility of using these two providers separately instead of setting them as a bundle.
@ZhenLian Do you have more thoughts on this?

Reply:
Yes, we wanted separate update intervals for root and identity cert, but after changing to ticker, I think we don't need two Providers anymore. We can use two tickers in the same for loop: https://stackoverflow.com/questions/40364270/running-a-function-periodically-in-go. @cindyxue what do you think? Can you make such changes?

Reply:
One quick question, we are still keeping `RootPEMFileProviderOptions` and `IdentityPEMFileProviderOptions` separately right?

Reply:
Done combining those :)